cvat-canvas/src/typescript/sliceHandler.ts

```diff
@@ -0,0 +1,589 @@
+
+export interface SliceHandler {
+    slice(sliceData: any): void;
+    transform(geometry: Geometry): void;
+    configurate(config: Configuration): void;
+    cancel(): void;
+}
+
+type EnhancedSliceData = {
+    enabled: boolean;
+    contour: number[];
+    state: any;
+    shapeType: 'mask' | 'polygon';
+};
+
+function drawOverOffscreenCanvas(context: OffscreenCanvasRenderingContext2D, image: CanvasImageSource): void {
+    context.fillStyle = 'black';
+    context.globalCompositeOperation = 'source-over';
+    context.drawImage(image, 0, 0);
+}
+
+function applyOffscreenCanvasMask(context: OffscreenCanvasRenderingContext2D, polygon: number[]): void {
+    const currentCompositeOperation = context.globalCompositeOperation;
+    context.globalCompositeOperation = 'destination-in';
+    context.beginPath();
+    context.moveTo(polygon[0], polygon[1]);
+    polygon.forEach((_, idx) => {
+        if (idx > 1 && !(idx % 2)) {
+            context.lineTo(polygon[idx], polygon[idx + 1]);
+        }
+    });
+    context.closePath();
+    context.fill();
+    context.globalCompositeOperation = currentCompositeOperation;
+}
+
+function indexGenerator(length: number, from: number, to: number, direction: 'forward' | 'backward'): number[] {
+    const result = [];
+    const value = direction === 'forward' ? 1 : -1;
+
+    if (from < 0 || from >= length || to < 0 || to >= length) {
+        throw new Error('Incorrect index generator input');
+    }
+
+    let i = from;
+    while (i !== to) {
+        result.push(i);
+        i += value;
+
+        if (i >= length) {
+            i = 0;
+        }
+
+        if (i < 0) {
+            i = length - 1;
+        }
+    }
+    result.push(i);
+    return result;
+}
+
+function getAllIntersections(segment: Segment, segments: Segment[]): Record<number, [number, number]> {
+    const intersections: Record<number, [number, number]> = {};
+    for (let i = 0; i < segments.length; i++) {
+        const checkedSegment = segments[i];
+        const intersection = findIntersection(checkedSegment, segment);
+        if (intersection !== null) {
+            intersections[i] = intersection;
+        }
+    }
+
+    return intersections;
+}
+
+export class SliceHandlerImpl implements SliceHandler {
+    private canvas: SVG.Container;
+    private startTimestamp: number;
+    private controlPointSize: number;
+    private outlinedBorders: string;
+    private enabled: boolean;
+    private shapeContour: SVG.PolyLine | null;
+    private slicingLine: SVG.PolyLine | null;
+    private slicingPoints: SVG.Circle[];
+    private hideObject: (clientID: number) => void;
+    private showObject: (clientID: number) => void;
+    private onSliceDone: (state?: any, results?: number[][], duration?: number) => void;
+    private onMessage: (messages: {
+        type: 'text' | 'list';
+        content: string | string[];
+        className?: string;
+        icon?: 'info' | 'loading';
+    }[] | null, topic: string) => void;
```

Comment:
I think we should create interface for message. Maybe we will use the tips somewhere else